### PR TITLE
Refactor as monorepo

### PR DESCRIPTION
**Pull request by mrcointreau**

### CODE REVIEW

File: assets/arc-0031/client/Dockerfile [hunk 1]
```diff
@@ -3,11 +3,15 @@ WORKDIR /client
...
+COPY client/package*.json ./client/
+COPY client/tsconfig.json ./client/
+RUN npm ci
 COPY . .
-RUN npm run build
+RUN npm run build --workspace=arc31 && npm run build --workspace=client
 
 FROM base AS production
-COPY --from=build /client/.output /client/.output
+COPY --from=build /client/client/.output ./.output
```
Comment (deanstef) [inline assets/arc-0031/client/Dockerfile]: What is going on with this doubled `client` dir?

File: assets/arc-0031/client/package.json [hunk 2]
```diff
@@ -1,9 +1,8 @@
 {
   "name": "arc-0031-client",
-  "version": "0.0.1",
   "type": "module",
-  "author": "Luca Martini",
-  "license": "UNLICENSED",
+  "version": "0.0.1",
+  "author": "[deanstef](https://github.com/deanstef) and [mrcointreau](https://github.com/mrcointreau)",
```
Comment (deanstef) [inline assets/arc-0031/client/package.json:5]: <3

File: assets/arc-0031/client/utils/hooks/useEnv.ts [hunk 3]
```diff
@@ -17,7 +19,11 @@ const createEnv = <TClient extends Record<string, ZodType>, TServer extends Reco
 
 export const useEnv = () => createEnv({
   client: {
-    algorandChainId: z.enum(['MainNet', 'TestNet', 'BetaNet'])
+    algorandChainId: (() => {
```
Comment (deanstef) [inline assets/arc-0031/client/utils/hooks/useEnv.ts]: Do we expect more chains apart from `MaiNet`, `TestNet` and `BetaNet`?

File: assets/arc-0031/server/src/mock/users.db.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import type { User } from '../types'
+
+const users: User[] = [
+  {
+    authAcc: 'REPLACE_WITH_ALGORAND_USER_ADDRESS',
+    challenge: '',
+    chainId: 'TestNet'
```
Comment (deanstef) [inline assets/arc-0031/server/src/mock/users.db.ts]: Here we should store the genesishas

File: assets/arc-0031/server/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import type { AlgorandChainId } from 'arc31'
+
+export interface User {
+  authAcc: string
+  challenge: string
+  chainId: AlgorandChainId
```
Comment (deanstef) [inline assets/arc-0031/server/src/types.ts]: As before, the chainId should be the genesis hash here

File: assets/arc-0031/client/utils/hooks/useEnv.ts (hunk 3 above)
Comment (mrcointreau) [reply assets/arc-0031/client/utils/hooks/useEnv.ts]: I don't think so, I'm going to restore the hardcoded enum values

File: assets/arc-0031/server/src/types.ts (hunk 5 above)
Comment (mrcointreau) [reply assets/arc-0031/server/src/types.ts]: Addressed in https://github.com/deanstef/ARCs/pull/4/commits/057fa7aa2cc37f11e5e0c4713e8a10f7f8bcd93f

File: assets/arc-0031/server/src/mock/users.db.ts (hunk 4 above)
Comment (mrcointreau) [reply assets/arc-0031/server/src/mock/users.db.ts]: Addressed in https://github.com/deanstef/ARCs/pull/4/commits/057fa7aa2cc37f11e5e0c4713e8a10f7f8bcd93f

File: assets/arc-0031/client/utils/hooks/useEnv.ts (hunk 3 above)
Comment (mrcointreau) [reply assets/arc-0031/client/utils/hooks/useEnv.ts]: Addressed in https://github.com/deanstef/ARCs/pull/4/commits/d7921dc99077fbd6ecab720e905be60bfe27d6e3

File: assets/arc-0031/server/src/types.ts (hunk 5 above)
Comment (mrcointreau) [reply assets/arc-0031/server/src/types.ts]: We are using `chainId` to refer to the Algorand chain networks (MainNet, TestNet and BetaNet) but in the `AuthMessage` interface we use it to refer to the chain network genesis hash, this is a little bit confusing.

File: assets/arc-0031/client/Dockerfile (hunk 1 above)
Comment (mrcointreau) [reply assets/arc-0031/client/Dockerfile]: In our Dockerfile, we are dealing with a multi-stage build process designed to handle our monorepo setup efficiently. The notation `client/client` that you see is a result of how we manage and build the client application within this Dockerized environment. Let me break it down for clarity:

```
FROM node:20-alpine AS base
WORKDIR /client
```

This sets up the base image and defines `/client` as our working directory. The `/client` directory here is at the root level of our Docker image's filesystem.

```
FROM base AS build
...
COPY client/package*.json ./client/
COPY client/tsconfig.json ./client/
...

```
In the build stage, we need to copy our local client directory's contents into the Docker image. We do this by copying them into `./client/`, which, relative to our current working directory `/client`, creates a directory structure like `/client/client/` inside the Docker image. The first client is our working directory, and the second client is the directory from our source code, which contains the actual client application code.

This step is crucial because it ensures that all necessary files for the client build (like package.json and tsconfig.json) are present in the Docker image.

```
FROM base AS production
COPY --from=build /client/client/.output ./.output
```

In the final production stage, we only need the build output of our client application. We copy this from `/client/client/.output` (which is where the built client resides in the image) to `./.output` in our production image. This ensures that the production image contains only the necessary built files, keeping it optimized and clean.

To summarize, the client/client directory structure in the Dockerfile is not an error but a necessary step in our build process. The first client refers to our Docker working directory, and the second client is the actual client application directory from our monorepo, being prepared and built in the Docker environment. This structure allows us to efficiently manage and build our client application within the Dockerized monorepo setup.

File: assets/arc-0031/server/src/types.ts (hunk 5 above)
Comment (deanstef) [reply assets/arc-0031/server/src/types.ts]: Good point. Let's clarify it and agree on the same terminology.

As defined in [ARC-6](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0006.md#string-specification-genesisid-and-genesishash)
- `GenesisID`: is the ASCII string of the Algorand network (e.g. `mainnet-v1.0`)
- `GenesisHash`: is the base64 encoded 32-byte hash of the genesis block

Whereas, as specified in [ARC-25](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0025.md#chain-ids):
- `ChainID`: is a standard numeric identifier of an Algorand chain.

I still think that we MUST use the GenesisID in the AuthMessage. It provides a unique link between the message and an Algorand chain, can be verified, and cannot change.

We can do that in another PR!

File: assets/arc-0031/client/Dockerfile (hunk 1 above)
Comment (deanstef) [reply assets/arc-0031/client/Dockerfile]: I see, thank you for taking the time to write such a precise clarification.

Can we call the `WORKDIR` differently then? Maybe we can have `WORKDIR /arc31` and thus `arc31/client/*` and `arc31/server` respectively for client and server Docker images. This would improve the readability and make it less confusing.

Comment (mrcointreau) [reply assets/arc-0031/client/Dockerfile]: Addressed in https://github.com/deanstef/ARCs/pull/4/commits/a803750a3ac2e78f1017220beafb53162d04b804

**Review by deanstef — APPROVED** [review]
